packages/vm/src/environment.rs

```diff
@@ -331,39 +349,29 @@ impl<A: BackendApi, S: Storage, Q: Querier> Environment<A, S, Q> {
     /// If the amount exceeds the available gas, the remaining gas is set to 0 and
     /// an VmError::GasDepletion error is returned.
     #[allow(unused)] // used in tests
-    pub fn decrease_gas_left(&self, amount: u64) -> VmResult<()> {
+    pub fn decrease_gas_left(&self, store: &mut impl AsStoreMut, amount: u64) -> VmResult<()> {
         self.with_wasmer_instance(|instance| {
-            let remaining = match get_remaining_points(instance) {
+            let remaining = match get_remaining_points(store, instance) {
                 MeteringPoints::Remaining(count) => count,
                 MeteringPoints::Exhausted => 0,
             };
             if amount > remaining {
-                set_remaining_points(instance, 0);
+                set_remaining_points(store, instance, 0);
                 Err(VmError::gas_depletion())
             } else {
-                set_remaining_points(instance, remaining - amount);
+                set_remaining_points(store, instance, remaining - amount);
                 Ok(())
             }
         })
     }
 
-    pub fn memory(&self) -> Memory {
-        self.with_wasmer_instance(|instance| {
-            let first: Option<Memory> = instance
-                .exports
-                .iter()
-                .memories()
-                .next()
-                .map(|pair| pair.1.clone());
-            // Every contract in CosmWasm must have exactly one exported memory.
-            // This is ensured by `check_wasm`/`check_wasm_memories`, which is called for every
-            // contract added to the Cache as well as in integration tests.
-            // It is possible to bypass this check when using `Instance::from_code` but then you
-            // learn the hard way when this panics, or when trying to upload the contract to chain.
-            let memory = first.expect("A contract must have exactly one exported memory.");
-            Ok(memory)
-        })
-        .expect("Wasmer instance is not set. This is a bug in the lifecycle.")
+    // Creates a MemoryView.
+    // This mut be short living and not be used after the memory was grown.
+    pub fn memory<'a>(&self, store: &'a mut impl AsStoreMut) -> MemoryView<'a> {
+        self.memory
+            .as_ref()
+            .expect("Memory is not set. This is a bug in the lifecycle.")
+            .view(store)
```

Comment:
Is this for performance or what? Just curious.

Reply:
It creates a MemoryView as designed by the new Wasmer API:

```rust
/// A WebAssembly `memory` view.
///
/// A memory view is used to read and write to the linear memory.
///
/// After a memory is grown a view must not be used anymore. Views are
/// created using the Memory.grow() method.
#[derive(Debug)]
pub struct MemoryView<'a>(pub(crate) memory_view_impl::MemoryView<'a>);
```

It seems like the `Memory` itself cannot be used directly:

```rust
#[derive(Debug, Clone, PartialEq)]
pub struct Memory(pub(crate) memory_impl::Memory);

impl Memory {
    // ...

    /// Creates a view into the memory that then allows for
    /// read and write
    pub fn view<'a>(&self, store: &'a impl AsStoreRef) -> MemoryView<'a> {
        MemoryView::new(self, store)
    }

    // ...
}
```